lib/tbot/service_spiffe_workload_api.go

```diff
@@ -0,0 +1,489 @@
+		s.log.WithFields(logrus.Fields{
+			"svid_type":     "x509",
+			"spiffe_id":     svidRes.SpiffeId,
+			"serial_number": serialString(cert.SerialNumber),
+			"hint":          svidRes.Hint,
+			"not_after":     cert.NotAfter,
+			"not_before":    cert.NotBefore,
+			"dns_sans":      cert.DNSNames,
+			"ip_sans":       cert.IPAddresses,
+			"serial":        cert.SerialNumber,
+		}).Info("Issued SVID for workload")
+	}
+
+	return svids, nil
+}
+
+// FetchX509SVID generates and returns the X.509 SVIDs available to a workload.
+// It is a streaming RPC, and sends renewed SVIDs to the client before they
+// expire.
+// Implements the SPIFFE Workload API FetchX509SVID method.
+func (s *SPIFFEWorkloadAPIService) FetchX509SVID(
+	_ *workloadpb.X509SVIDRequest,
+	srv workloadpb.SpiffeWorkloadAPI_FetchX509SVIDServer,
+) error {
+	renewCh, unsubscribe := s.trustBundleBroadcast.subscribe()
+	defer unsubscribe()
+	s.log.Info("FetchX509SVID stream opened by workload")
+	defer s.log.Info("FetchX509SVID stream has closed")
+
+	for {
+		s.log.Info("Starting to issue X509 SVIDs to workload")
+
+		svids, err := s.fetchX509SVIDs(srv.Context())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		err = srv.Send(&workloadpb.X509SVIDResponse{
+			Svids: svids,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		s.log.Debug(
+			"Finished issuing SVIDs to workload. Waiting for next renewal interval or CA rotation",
+		)
+
+		select {
+		case <-srv.Context().Done():
+			s.log.Debug("Context closed, stopping SVID stream")
+			return nil
+		case <-time.After(s.botCfg.RenewalInterval):
+			s.log.Debug("Renewal interval reached, renewing SVIDs")
+			// Time to renew the certificate
+			continue
+		case <-renewCh:
+			s.log.Debug("Trust bundle has been updated, renewing SVIDs")
+			continue
+		}
+	}
+}
+
+// FetchX509Bundles returns the trust bundle for the trust domain. It is a
+// streaming RPC, and will send rotated trust bundles to the client for as long
+// as the client is connected.
+// Implements the SPIFFE Workload API FetchX509SVID method.
+func (s *SPIFFEWorkloadAPIService) FetchX509Bundles(
+	_ *workloadpb.X509BundlesRequest,
+	srv workloadpb.SpiffeWorkloadAPI_FetchX509BundlesServer,
+) error {
+	s.log.Info("FetchX509Bundles stream opened by workload")
+	defer s.log.Info("FetchX509Bundles stream has closed")
+	renewCh, unsubscribe := s.trustBundleBroadcast.subscribe()
+	defer unsubscribe()
+
+	for {
+		s.log.Info("Sending X.509 trust bundles to workload")
+		err := srv.Send(&workloadpb.X509BundlesResponse{
+			// Bundles keyed by trust domain
+			Bundles: map[string][]byte{
+				s.trustDomain: s.getTrustBundle(),
+			},
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		select {
+		case <-srv.Context().Done():
+			s.log.Debug("Context closed, stopping x.509 trust bundle stream")
+			return nil
+		case <-renewCh:
+			s.log.Debug("Trust bundle has been updated, resending trust bundle")
+			continue
+		}
+	}
```

Comment:
Just double checking: no need to listen for cert rotations here?

Reply:
As per https://github.com/gravitational/teleport/pull/37772/files#r1512389858